File: rust/cmd/doubaospeech/src/commands/realtime.rs
//! Realtime (Real-time voice conversation) commands.

use std::io::Write;
use std::time::Duration;

use clap::{Args, Subcommand};
use serde::{Deserialize, Serialize};
use tokio::time::timeout;

use giztoy_doubaospeech::{
    RealtimeASRConfig, RealtimeAudioConfig, RealtimeConfig, RealtimeDialogConfig,
    RealtimeEventType, RealtimeTTSConfig,
};

use super::{create_client, get_context, load_request, output_result, print_success, print_verbose};
use crate::Cli;

/// Realtime voice conversation service.
///
/// Enables bidirectional voice communication with AI.
#[derive(Args)]
pub struct RealtimeCommand {
    #[command(subcommand)]
    command: RealtimeSubcommand,
}

#[derive(Subcommand)]
enum RealtimeSubcommand {
    /// Connect to realtime service
    Connect,
    /// Test realtime connection (send greeting and receive response)
    Test {
        /// Greeting text to send
        #[arg(short = 'g', long, default_value = "你好")]
        greeting: String,
    },
}

/// Realtime config from YAML/JSON file.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct RealtimeFileConfig {
    #[serde(default)]
    asr: AsrConfig,
    #[serde(default)]
    tts: TtsConfig,
    #[serde(default)]
    dialog: DialogConfig,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct AsrConfig {
    #[serde(default)]
    extra: std::collections::HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct TtsConfig {
    #[serde(default)]
    speaker: String,
    #[serde(default)]
    audio_config: AudioConfig,
    #[serde(default)]
    extra: std::collections::HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct AudioConfig {
    #[serde(default = "default_channel")]
    channel: i32,
    #[serde(default = "default_format")]
    format: String,
    #[serde(default = "default_sample_rate")]
    sample_rate: i32,
}

impl Default for AudioConfig {
    fn default() -> Self {
        Self {
            channel: 1,
            format: "mp3".to_string(),
            sample_rate: 24000,
        }
    }
}

fn default_channel() -> i32 {
    1
}
fn default_format() -> String {
    "mp3".to_string()
}
fn default_sample_rate() -> i32 {
    24000
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
struct DialogConfig {
    #[serde(default)]
    bot_name: String,
    #[serde(default)]
    system_role: String,
    #[serde(default)]
    speaking_style: String,
    #[serde(default)]
    character_manifest: String,
    #[serde(default)]
    extra: std::collections::HashMap<String, serde_json::Value>,
}

impl RealtimeFileConfig {
    fn to_realtime_config(&self) -> RealtimeConfig {
        RealtimeConfig {
            asr: RealtimeASRConfig {
                extra: self.asr.extra.clone(),
            },
            tts: RealtimeTTSConfig {
                speaker: self.tts.speaker.clone(),
                audio_config: RealtimeAudioConfig {
                    channel: self.tts.audio_config.channel,
                    format: self.tts.audio_config.format.clone(),
                    sample_rate: self.tts.audio_config.sample_rate,
                },
                extra: self.tts.extra.clone(),
            },
            dialog: RealtimeDialogConfig {
                bot_name: self.dialog.bot_name.clone(),
                system_role: self.dialog.system_role.clone(),
                speaking_style: self.dialog.speaking_style.clone(),
                character_manifest: self.dialog.character_manifest.clone(),
                extra: self.dialog.extra.clone(),
                ..Default::default()
            },
        }
    }
}

impl RealtimeCommand {
    pub async fn run(&self, cli: &Cli) -> anyhow::Result<()> {
        match &self.command {
            RealtimeSubcommand::Connect => self.connect(cli).await,
            RealtimeSubcommand::Test { greeting } => self.test(cli, greeting).await,
        }
    }

    async fn connect(&self, cli: &Cli) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        let client = create_client(&ctx)?;

        // Load config from file
        let config = if let Some(input_file) = cli.input.as_deref() {
            let file_config: RealtimeFileConfig = load_request(input_file)?;
            file_config.to_realtime_config()
        } else {
            // Default config
            RealtimeConfig {
                tts: RealtimeTTSConfig {
                    speaker: "zh_female_cancan".to_string(),
                    audio_config: RealtimeAudioConfig::default(),
                    ..Default::default()
                },
                dialog: RealtimeDialogConfig {
                    bot_name: "助手".to_string(),
                    system_role: "你是一个友好的语音助手".to_string(),
                    ..Default::default()
                },
                ..Default::default()
            }
        };

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(cli, &format!("Speaker: {}", config.tts.speaker));

        // Connect to realtime service
        eprintln!("Connecting to realtime service...");
        let session = client.realtime().connect(&config).await?;

        print_success(&format!("Connected! Session ID: {}", session.session_id()));

        // Keep connection alive and print events
        eprintln!("Listening for events... (Press Ctrl+C to exit)");

        loop {
            match timeout(Duration::from_secs(30), session.recv()).await {
                Ok(Some(Ok(event))) => {
                    match event.event_type {
                        Some(RealtimeEventType::AudioReceived) => {
                            if let Some(ref audio) = event.audio {
                                eprint!("🔊 Audio received: {} bytes\r", audio.len());
                                std::io::stderr().flush().ok();
                            }
                        }
                        Some(RealtimeEventType::ASRFinished) => {
                            if let Some(ref info) = event.asr_info {
                                eprintln!("🎤 ASR: {}", info.text);
                            }
                        }
                        Some(RealtimeEventType::TTSStarted) => {
                            eprintln!("🗣️ TTS: {}", event.text);
                        }
                        Some(RealtimeEventType::SessionFailed) => {
                            if let Some(ref err) = event.error {
                                eprintln!("❌ Session failed: {}", err);
                            }
                            break;
                        }
                        Some(RealtimeEventType::SessionEnded) => {
                            eprintln!("Session ended");
                            break;
                        }
                        _ => {
                            print_verbose(cli, &format!("Event: {:?}", event.event_type));
                        }
                    }
                }
                Ok(Some(Err(e))) => {
                    eprintln!("Error: {}", e);
                    break;
                }
                Ok(None) => {
                    eprintln!("Connection closed");
                    break;
                }
                Err(_) => {
                    // Timeout, keep waiting
                    continue;
                }
            }
        }

        session.close().await?;
        Ok(())
    }

    async fn test(&self, cli: &Cli, greeting: &str) -> anyhow::Result<()> {
        let ctx = get_context(cli)?;
        let client = create_client(&ctx)?;

        // Load config from file or use default
        let config = if let Some(input_file) = cli.input.as_deref() {
            let file_config: RealtimeFileConfig = load_request(input_file)?;
            file_config.to_realtime_config()
        } else {
            RealtimeConfig {
                tts: RealtimeTTSConfig {
                    speaker: "zh_female_cancan".to_string(),
                    audio_config: RealtimeAudioConfig::default(),
                    ..Default::default()
                },
                dialog: RealtimeDialogConfig {
                    bot_name: "助手".to_string(),
                    system_role: "你是一个友好的语音助手".to_string(),
                    ..Default::default()
                },
                ..Default::default()
            }
        };

        print_verbose(cli, &format!("Using context: {}", ctx.name));

        // Connect
        eprintln!("Connecting to realtime service...");
        let session = client.realtime().connect(&config).await?;
        print_success(&format!("Connected! Session ID: {}", session.session_id()));

        // Send greeting
        eprintln!("Sending greeting: {}", greeting);
        session.say_hello(greeting).await?;

        // Collect responses
        let mut audio_bytes = 0usize;
        let mut response_text = String::new();

        eprintln!("Waiting for response...");

        loop {
            match timeout(Duration::from_secs(10), session.recv()).await {
                Ok(Some(Ok(event))) => {
                    match event.event_type {
                        Some(RealtimeEventType::AudioReceived) => {
                            if let Some(ref audio) = event.audio {
                                audio_bytes += audio.len();
                            }
                        }
                        Some(RealtimeEventType::TTSStarted) => {
                            response_text = event.text.clone();
                            eprintln!("🗣️ Response: {}", event.text);
                        }
                        Some(RealtimeEventType::TTSFinished) => {
                            break;
                        }
                        Some(RealtimeEventType::SessionFailed) => {
                            if let Some(ref err) = event.error {
                                return Err(anyhow::anyhow!("Session failed: {}", err));
                            }
                            break;
                        }
                        _ => {}
                    }
                }
                Ok(Some(Err(e))) => {
                    return Err(anyhow::anyhow!("Error: {}", e));
                }
                Ok(None) => {
                    break;
                }
                Err(_) => {
                    // Timeout
                    break;
                }
            }
        }

        session.close().await?;

        let result = serde_json::json!({
            "session_id": session.session_id(),
            "greeting": greeting,
            "response_text": response_text,
            "audio_bytes": audio_bytes,
        });

        output_result(&result, cli.output.as_deref(), cli.json)
    }
}

File: go/cmd/doubaospeech/commands/realtime.go
package commands

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"os"
	"time"

	"github.com/spf13/cobra"

	ds "github.com/haivivi/giztoy/go/pkg/doubaospeech"
)

var realtimeCmd = &cobra.Command{
	Use:   "realtime",
	Short: "Real-time voice conversation service",
	Long: `Real-time end-to-end voice conversation service.

Enables bidirectional voice communication with AI.

Example config file (realtime.yaml):
  asr:
    extra:
      end_smooth_window_ms: 200
  tts:
    speaker: zh_female_cancan
    audio_config:
      channel: 1
      format: pcm
      sample_rate: 24000
  dialog:
    bot_name: 小助手
    system_role: 你是一个友好的语音助手
    speaking_style: 温柔亲切`,
}

var realtimeConnectCmd = &cobra.Command{
	Use:   "connect",
	Short: "Connect to realtime service",
	Long: `Connect to the real-time conversation service.

Establishes a WebSocket connection for bidirectional communication.

Examples:
  doubaospeech -c myctx realtime connect -f realtime.yaml`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		cliCtx, err := getContext()
		if err != nil {
			return err
		}

		client, err := createClient(cliCtx)
		if err != nil {
			return err
		}

		var config ds.RealtimeConfig
		if err := loadRequest(getInputFile(), &config); err != nil {
			return err
		}

		printVerbose("Using context: %s", cliCtx.Name)

		audioFile, _ := cmd.Flags().GetString("audio")

		reqCtx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
		defer cancel()

		return runRealtimeConnect(reqCtx, client, &config, audioFile)
	},
}

var realtimeInteractiveCmd = &cobra.Command{
	Use:   "interactive",
	Short: "Interactive voice conversation",
	Long: `Start an interactive voice conversation session.

This mode captures audio from your microphone and plays
responses through your speakers.

Examples:
  doubaospeech -c myctx realtime interactive -f realtime.yaml`,
	RunE: func(cmd *cobra.Command, args []string) error {
		return fmt.Errorf("not implemented: interactive realtime mode requires microphone/speaker hardware integration; use 'doubaospeech realtime connect' for programmatic access")
	},
}

// ============================================================================
// Implementation Functions
// ============================================================================

func runRealtimeConnect(ctx context.Context, client *ds.Client, config *ds.RealtimeConfig, audioFile string) error {
	session, err := client.Realtime.Connect(ctx, config)
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}
	defer session.Close()

	printSuccess("Connected to realtime service (session: %s)", session.SessionID())

	// If audio file provided, send it; otherwise read from stdin
	if audioFile != "" {
		audioData, err := os.ReadFile(audioFile)
		if err != nil {
			return fmt.Errorf("read audio file: %w", err)
		}

		printVerbose("Sending audio (%s)...", formatBytes(int64(len(audioData))))

		// Send audio in chunks
		chunkSize := 3200 // 100ms of 16kHz 16-bit mono
		for i := 0; i < len(audioData); i += chunkSize {
			end := i + chunkSize
			if end > len(audioData) {
				end = len(audioData)
			}

			if err := session.SendAudio(ctx, audioData[i:end]); err != nil {
				return fmt.Errorf("send audio: %w", err)
			}
			time.Sleep(100 * time.Millisecond)
		}
	} else {
		// Read audio from stdin
		var buf bytes.Buffer
		if _, err := io.Copy(&buf, os.Stdin); err != nil {
			return fmt.Errorf("read stdin: %w", err)
		}
		audioData := buf.Bytes()

		if len(audioData) > 0 {
			printVerbose("Sending audio from stdin (%s)...", formatBytes(int64(len(audioData))))
			chunkSize := 3200
			for i := 0; i < len(audioData); i += chunkSize {
				end := i + chunkSize
				if end > len(audioData) {
					end = len(audioData)
				}
				if err := session.SendAudio(ctx, audioData[i:end]); err != nil {
					return fmt.Errorf("send audio: %w", err)
				}
				time.Sleep(100 * time.Millisecond)
			}
		}
	}

	// Receive events
	outputPath := getOutputFile()
	var audioBuf bytes.Buffer

	for event, err := range session.Recv() {
		if err != nil {
			return fmt.Errorf("receive error: %w", err)
		}

		switch event.Type {
		case ds.EventASRResponse:
			if event.ASRInfo != nil {
				if event.ASRInfo.IsFinal {
					printInfo("[ASR] %s", event.ASRInfo.Text)
				} else {
					printVerbose("[ASR interim] %s", event.ASRInfo.Text)
				}
			}
		case ds.EventChatResponse:
			fmt.Print(event.Text)
		case ds.EventChatEnded:
			fmt.Println()
		case ds.EventAudioReceived:
			if len(event.Audio) > 0 {
				audioBuf.Write(event.Audio)
				printVerbose("[Audio] +%d bytes", len(event.Audio))
			}
		case ds.EventSessionFailed:
			if event.Error != nil {
				return fmt.Errorf("session failed: %s", event.Error.Message)
			}
		case ds.EventSessionFinished:
			printVerbose("Session finished")
			goto done
		}
	}

done:
	if audioBuf.Len() > 0 && outputPath != "" {
		if err := outputBytes(audioBuf.Bytes(), outputPath); err != nil {
			return fmt.Errorf("write audio: %w", err)
		}
		printSuccess("Audio saved to: %s (%s)", outputPath, formatBytes(int64(audioBuf.Len())))
	}

	return nil
}

func init() {
	realtimeConnectCmd.Flags().String("audio", "", "Audio file to send (optional, defaults to stdin)")

	realtimeCmd.AddCommand(realtimeConnectCmd)
	realtimeCmd.AddCommand(realtimeInteractiveCmd)
}
